src/stream/src/executor/dispatch.rs

```diff
@@ -271,7 +277,10 @@ impl StreamConsumer for DispatchExecutor {
                     })
                     .await?;
                 if let Some(barrier) = barrier {
-                    yield barrier;
+                    yield ActorTrapItem::Barrier(barrier);
+                }
+                while let Some(msg) = self.control_msg_rx.try_recv()? {
```

Comment:
This might not work in the dispatcher if this actor has stuck and no message can be handled. How about directly let the actor to poll it by `select`ing `Message` and `ControlMessage`?

Reply:
Furthermore, we can even select the `actor.run()`.

```rust
tokio::select! {
  _ = actor.run(),
  _ = async {
    while let Some(msg) = rx.recv().await {
       self.env.state_store().apply_control_msg(msg).await;
    }
  }
}
```

Reply:
I used to consider this implementation. However, in this way we cannot ensure that no executor is processing a stream message when the we receive a control message from the channel. 

For example, while an executor is reading from the state store, it may stuck in an await point, such as fetching data from remote object store. The `actor.run().poll()` will return a `Pending`. And then we may have a control message in the channel and the `select!` will take the message out and later apply it to state store. However, the executor are still pending on reading from the state store, and we should not modify the state store, or otherwise something wrong may happen.

Reply:
> However, in this way we cannot ensure that no executor is processing a stream message when the we receive a control message from the channel.

It's still possible, considering barrier align of Join.

However, I believe the message applying is designed to be without contending, that is, accessing and updating the internal structure of the local state store can always be done in a synchronous manner. So there's no yield point during this procedure, and no need to be concerned about it. 🤔

Reply:
Applying the message can indeed be done in a synchronous manner. The problem lies in that the executor is accessing the internal structure in a async manner.

For example, we have an async method look like the following
```
async fn process<'a>(local_state_store: &'a LocalStateStore) {
    local_state_store.do_some_io().await;
}
```

And we have a `select!`
```
let local_state_store = Arc::new(Mutex::new(...))

tokio::select! {
  _ = async move {
        local_state_store.try_lock().unwrap().process().await;
    },
  _ = async {
    while let Some(msg) = rx.recv().await {
       local_state_store.try_lock().unwrap().apply_control_msg(msg);
    }
  }
}
```

While the `local_state_store.process()` is pending on `do_some_io()`, it will not release the lock acquired from `try_lock`. `local_state_store.process()` is pending and will hand over the execution to `select`, and then `select` will poll from `rx.recv()`, and if it gets a control message, it will call `try_lock` and fail because the lock is still held by `local_state_store.process()`. This might be the case mentioned by @MrCroxx when we discussed on Slack.

But it seems that if we use `tokio::sync::Mutex`, and we use `lock().await` instead of `try_lock`, there will not be such problem.

Reply:
I thought there was no async operation in `process`... 🥵 do we really need to do some I/O in it?

> and we use lock().await instead of try_lock, there will not be such problem.

Then we'll not even ensure that "no executor is processing a stream message". :(

Reply:
If `apply_control_msg` (i.e., modifying some structure) can be done synchronously, it means that using this structure must be done synchronously too. We may use a synchronous lock here, so we don't need to worry about yielding when using. 🤔

Reply:
> I thought there was no async operation in process... 🥵 do we really need to do some I/O in it?

The `process` can be `get` and `iter`. We need to download the SST data and meta, so I/O is unavoidable and there will be await point.

> We may use a synchronous lock here

We cannot use synchronous lock here, since if we hold a lock guard across the await point, the future will become `!Send + !Sync` since the future will have to hold the lock guard, which is not `Send + Sync`.

The code below will get a compile error
```
async fn some_async() {}

pub async fn async_test() {
    let lock = parking_lot::Mutex::new(1);
    let guard = lock.lock();
    some_async().await;
    assert_eq!(*guard, 1);
}

pub fn spawn_task() {
    tokio::spawn(async_test());
}
```

```
error: future cannot be sent between threads safely
   --> src/storage/src/hummock/iterator/mod.rs:350:18
    |
350 |     tokio::spawn(async_test());
    |                  ^^^^^^^^^^^^ future returned by `async_test` is not `Send`
    |
    = help: within `impl futures::Future<Output = ()>`, the trait `std::marker::Send` is not implemented for `*mut ()`
note: future is not `Send` as this value is used across an await
   --> src/storage/src/hummock/iterator/mod.rs:345:17
    |
344 |     let guard = lock.lock();
    |         ----- has type `parking_lot::lock_api::MutexGuard<'_, parking_lot::RawMutex, i32>` which is not `Send`
345 |     some_async().await;
    |                 ^^^^^^ await occurs here, with `guard` maybe used later
346 |     assert_eq!(*guard, 1);
347 | }

```

Reply:
If we use `tokio::sync::Mutex` and use `lock().await`, we can solve this problem, and we can do `select actor.run() + rx.recv().await`